readme
/******************************************************************************
 *
 *  Copyright 2018-2022 NXP
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at:
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 ******************************************************************************/

libbt supports to configure some key parameters with conf file in default path /vendor/etc/bluetooth/bt_vendor.conf

use "key = value" to set parameters in conf file

optional keys are listed below:

	mchar_port: port path used by Uart interface, the default BT uart port in libbt is /dev/ttyUSB0 
                          if it's not the same as default port, could set mchar_port in conf file like below
                          mchar_port = /dev/ttyXXX

	mbt_port: port path used by SD or USB interface, the default BT SD/USB port in libbt is /dev/mbtchar0
	               if it's not the same as default port, could set mbt_port in conf file like below
		        mbt_port = /dev/mbtcharX

	baudrate_fw_init: default bardrate when bluetooth fw active after download. The default value is 115200 in libbt. It's not necessary to configure this key in conf unless it need to choose other baudrate.

	baudrate_bt: the baudrate expect to working on when bt running. The default value is 3000000 in libbt. It's not necessary to configure this key in conf unless it need to choose other baudrate. 
	
	bd_address: set bdaddress by configuration file, bdaddress is comprised by 6 bytes. You are expected to input as the format of "bd_address = xx:xx:xx:xx:xx:xx", where 'x' should be hexadecimal digits.

	use_controller_addr: Use controller address if its not default and bd_address is not used.
				Supported values:
				use_controller_addr = 1 (Use controller address, default)
				use_controller_addr = 0 (Dont use controller address)

	pFilename_cal_data: bt calibration file, example: pFilename_cal_data = /vendor/firmware/BtCalData_ext_Boreal_P0_V2.conf
						Note: The Baud rate within calibration file should be same as baudrate_bt(default value is 3000000) in bt_vendor.conf file.

	ble_1m_power: Set signed Bluetooth LE Transmit Power level for 1M PHY, example: ble_1m_power = 10
	ble_2m_power: Set signed Bluetooth LE Transmit Power level for 2M PHY, example: ble_1m_power = -5
				Valid Supported Values:
						Power Class		Minimum Tx Power 	Maximum Tx Power 
							1.5				 -20dbm				 10dbm
							2				 -20dbm				 4 dbm
				Note: Incase the paramter is not set, firmware default values will be used.

	bt_max_power_sel: Update classic BT max transmit power level.
				Supported values:
					bt_max_power_sel = 0 (Use firmware default max power level)
					bt_max_power_sel = 1 (Initial power provided by calibration file will be used as reference for Max Tx power)

				Note: This flag updates the maximum transmit power level.

	pFilename_fw_init_config_bin: Provide path of FW initialization configuration binary.

				Example: pFilename_fw_init_config_bin = /vendor/firmware/88W8987A0_FwConf_0.bin

				Note: Make sure bt_vendor.conf paramter are updated based on feature in pFilename_fw_init_config_bin file.
					  Feature is supported only when BT only firmware download is used.
					  Refer conf/bin for more info.

	independent_reset_mode: Select independent reset modes. 
				Supported values:

					independent_reset_mode = 0 (Disable IR configuration, default)

					independent_reset_mode = 1 (Out Of Band Independent Reset during FW configuration)
						Make sure send_oob_ir_trigger paramter is configured in bt_vendor.conf

					independent_reset_mode = 2 (Inband Independent Reset during FW configuration)

	send_oob_ir_trigger : Configure the Independent Reset(IR) Trigger method.
				Supported values:
				send_oob_ir_trigger = 0 (Disable IR Trigger, default)

				send_oob_ir_trigger = 1 (Use RFKILL to trigger IR)

				send_oob_ir_trigger = 2 (Use GPIO to trigger IR)
					Note: Incase using GPIO trigger make sure independent_reset_gpio_pin, oob_ir_host_gpio_pin, chardev_name are configured in bt_vendor.conf

	independent_reset_gpio_pin: SOC specific GPIO pins selected for triggering out-band reset
				Example: independent_reset_gpio_pin = 15(default value is 255(0xFF))

    oob_ir_host_gpio_pin: Exposed host specific GPIO pin on i.MX platform used to trigger pulse for IR
				Example:ir_host_gpio_pin = 14(For i.MX8 Mini).

    chardev_name: Open the desired GPIO chip by calling gpiod_chip_open API 
				Example:chardev_name="/dev/gpiochip5" by default(For i.MX8 Mini).

	enable_sco_config: Configure sco config.
				Supported values:
				enable_sco_config = 1 (Enable SCO cofig, default)
				enable_sco_config = 0 (Disable SCO config)

	enable_pdn_recovery: Enable PDn recovery mechanism support.
				Supported values:
				enable_pdn_recovery = 0 (Disable, default)
				enable_pdn_recovery = 1 (Enable)

	vhal_trace_level : Select log level for BT VHAL module. Note each level will include messages from previous level as well.
				Supported values:
				vhal_trace_level = 0    No trace messages
				vhal_trace_level = 1    Error condition trace messages
				vhal_trace_level = 2    Warning condition trace messages
				vhal_trace_level = 3    Generic Info messages    
				vhal_trace_level = 4    Debug messages
				vhal_trace_level = 5    Verbose trace message

	enable_lpm : Configure low power mode (Host to Controller Sleep (H2C))
				Supported values:
				enable_lpm = 0    (Disable, Default)
				enable_lpm = 1    (Enable)

	lpm_timeout_ms : Set host timeout in milliseconds after which controller is triggered to sleep.
				Example:
				lpm_timeout_ms = 300 (Default value is 1000 ms)
					Note: Make sure enable_lpm is enabled to use this configuration.

Below parameters are for fw download, if not use fw download by libbt, don't set any of below in conf file

	enable_download_fw: set to 1 if need to download uart bt fw by libbt when bootup, default value is 0 in libbt, it always download combo fw by wifi side.

	enable_poke_controller: Support for faster firmware download
                        Supported values:
                        enable_poke_controller = 0 (default)
                        enable_poke_controller = 1 (for IW512, IW611, IW612,AW611HN/AOA,AW611HN/AOB,9098 chipsets)

	pFileName_image: Bt fw path.

		Example: pFileName_image = /vendor/firmware/uart8997_bt_v4.bin

		Note: Setting this value can be skipped for following SOC, respective default values will be used:

				9098 A1 and A2 chip 			: /vendor/firmware/uart9098_bt_v1.bin
				IW512/AW611/IW611/IW612 A0 chip	: /vendor/firmware/uartspi_n61x.bin
				IW512/AW611/IW611/IW612 A1 chip : /vendor/firmware/uartspi_n61x_v1.bin

	pFileName_helper: bt fw helper path, example: pFileName_helper = /vendor/firmware/helper_uart_3000000.bin

	iSecondBaudrate: second baudrate used when download fw. The default value is 0 in libbt, only need to configure it if for 90xx chips.
	example: iSecondBaudrate = 3000000

	enable_heartbeat_config: Used to enable/disable HEARTBEAT Configurations.
				Supported Values:
				enable_heartbeat_config = 0 (disable, Default)
				enable_heartbeat_config = 1 (enable)

	wakeup_power_gpio_pin : GPIO pin selected to generate high/low interrupt to wake up host when receiving the specific advertising packet from RCU Power key
		default value: 13

	wakeup_power_gpio_high_duration : the time interval for making the GPIO pin in high state for receiving the specific advertising packet from RCU Power key, unit is 100ms
		default value: 2

	wakeup_power_gpio_low_duration : the time interval for making the GPIO pin in low state for receiving the specific advertising packet from RCU Power key, unit is 100ms
		default value : 2

	wakeup_netflix_gpio_pin : GPIO pin selected to generate high/low interrupt to wake up host when receiving the specific advertising packet from RCU Netflix key
		default value: 13

	wakeup_netflix_gpio_high_duration : the time interval for making the GPIO pin in low state for receiving the specific advertising packet from RCU Netflix key, unit is 100ms
		default value : 4

	wakeup_netflix_gpio_low_duration : the time interval for making the GPIO pin in low state for receiving the specific advertising packet from RCU Netflix key, unit is 100ms
		default value : 4

	wakeup_adv_pattern : the pattern value to compare with the AD structure value in advertising packet from the air, if all the value matches perfectly, firmware will toggle the GPIO to wake up host, which is in sleep state, based on the configurations, maximum length is 16 bytes
		Configuration rule : the order of value matters, so the configuration value shall be put in the correct sequence
		Example: the pattern value for case 1 will be 0x00, 0x01 ,0x02, 0x03
				 the pattern value for case 2 will be 0x03, 0x02 ,0x01, 0x00
			case 1:
				wakeup_adv_pattern = 0
				wakeup_adv_pattern = 1
				wakeup_adv_pattern = 2
				wakeup_adv_pattern = 3
			case 2:
				wakeup_adv_pattern = 3
				wakeup_adv_pattern = 2
				wakeup_adv_pattern = 1
				wakeup_adv_pattern = 0
		default value : NONE

	wakeup_scan_type : same definition with legacy scan parameter in Core Spec(7.8.10 in v5.3), used to wakeup scan after host entering sleep state
		default value : 0

	wakeup_scan_interval : same definition with legacy scan parameter in Core Spec(section 7.8.10 in v5.3), used to wakeup scan after host entering sleep state
		default value : 128

	wakeup_scan_window : same definition with legacy scan parameter in Core Spec(section 7.8.10 in v5.3), used to wakeup scan after host entering sleep state
		default value : 96

	wakeup_own_addr_type : same definition with legacy scan parameter in Core Spec(section 7.8.10 in v5.3), used to wakeup scan after host entering sleep state
		default value : 0

	wakeup_scan_filter_policy : same definition with legacy scan parameter in Core Spec(section 7.8.10 in v5.3), used to wakeup scan after host entering the sleep state
		default value : 0

	wakeup_local_heartbeat_timer_value : the time interval for HAL to periodically send a specific VSC, called heartbeat command, to firmware, to notify that host is still in the wake state, unit is 100ms
		default value : 8

	wakeup_enable_uart_low_config: Configure controller to pull controller specific UART lines to low when HEARTBEAT timer is timed out on controller.
				Supported Values:
				wakeup_enable_uart_low_config = 0 (disable, Default)
				wakeup_enable_uart_low_config = 1 (enable)

       Below parameters are not necessary to configure in conf file when download fw by libbt.

	baudrate_dl_helper: baudrate used when download helper. The default value is 115200 in libbt, only need to configure it if choose other baudrate. 

	baudrate_dl_image: baudrate used when start download fw. The default value is 3000000 in libbt, only need to configure it if choose other baudrate.  

	uart_sleep_after_dl: sleep time after bt fw is downloaded, unit in ms. The default value is 100 ms, only need to configure it if need sleep for more time.

	send_boot_sleep_trigger: Used to send trigger from Boot Sleep Patch
		Supported Values:
		send_boot_sleep_trigger = 0 (Default)
		send_boot_sleep_trigger = 1 (Enable sending  Boot Sleep Patch Trigger)


CHANGELOG.md
# Change Log
All notable changes in Bluetooth Vendor Hardware Abstraction Layer(BT VHAL) will be documented in this file.

## [009.022] - 23-October-20

### Fixed
- Fixed Code Sonar warnings

## [009.021] - 23-September-01

### Fixed
- Correction of typos

## [009.019] - 2023-July-27

### Fixed
- Added sign compare in build flag and fixed related issues

## [009.017] - 2023-June-13

### Added
- Updated code to send exit heartbeat mode command to controller on BT off event
- Added wakeup_enable_uart_low_config parameter to pull controller specific UART lines to low when HEARTBEAT timer is timed out on controller.
- Updated code in BT off event to avoid unexpected FW condition and data on BT enable event

## [009.016] - 2023-March-24

### Fixed
- Fixed static code analysis errors

## [009.015] - 2023-March-24

### Added
- Added Low Power Mode(LPM) support via Inband method

## [009.013] - 2022-Oct-13

### Fixed
- Fixed all coverity issues.

## [009.012] - 2022-Sep-30

### Fixed
- Updated Makefile to fix BT VHAL compilation error for Android-13 codebase

## [009.011] - 2022-Sep-22

### Added
- Added Boot Sleep Patch support for Android Auto
 
## [009.010] - 2022-Aug-26

### Fixed
- Fixed Inband Independent Reset Trigger for Android Automotive

## [009.008] - 2022-July-15

### Added
- Code updated to use unanimous function for logging and support 5 log level.

## [009.006] - 2022-Jun-03

### Fixed
- Remove extra wait if Firmware is already downloaded to reduce BT turn on time.

## [009.005] - 2022-Apr-21

### Fixed
- Removed compile time flag wrt BLE wake on feature for OTT project.

## [009.003] - 2022-Apr-01

### Added
- Automatically select firmware name from BT VHAL for Chips supporting boot code version 3

## [009.000] - 2022-Feb-04
 
### Added
- Remove compile time feature flag and make it runtime configurable.

### Fixed
- Add Poke command to speed up Firmware download for latest Chipset.

## [008.018] - 2022-Jan-07
 
### Fixed
- Handle "Non-empty else"" log print, by cleaning extra message in queue.
  
## [008.015] - 2021-Oct-20
 
### Added
- BLE wake on feature for OTT project

## [008.014] - 2021-Oct-14
 
### Added
- Inband Independent Reset in BT Vendor HAL.

## [008.013] - 2021-Sep-17

### Fixed
- Added FIFO check before reading data to avoid extra time for Change Buadrate Command.

## [008.011] - 2021-Aug-01
 
### Added
- Support user space host GPIO toggling from VHAL for triggering Out of band Independent Reset om imx platform.

## [008.010] - 2021-Jul-30
 
### Added
- Enable BT Tx Power Set capability in VHAL

## [008.009] - 2021-Jun-24
 
### Fixed
- Avoid going in infinite loop incase bootloader signature is not received during stress test.
- Handle FD leak causing BT crash during stress test.
- Added Delay after Change Buadrate to provide sufficient time for command to complete.

## [008.007] - 2021-May-19
 
### Added
- Support Out of Band(OOB) Independent Reset(IR) in vendor HAL
- Integration of RFKILL patch to toggle GPIO for OOB IR
- Print BT firmware version from vendor HAL

## [008.005] - 2021-May-18

### Added
 - Set BLE TX power level through vendor HAL for 1M and 2M PHY

## [008.004] - 2021-Apr-26

### Added
 - Upgraded Firmware Loader version to M322.

## [008.003] - 2021-Apr-08
 
### Added
- Loading BT Calibration Data via BT Vendor HAL


bt_vendor_nxp.h
/******************************************************************************
 *
 *  Copyright 2018-2023 NXP
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at:
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 ******************************************************************************/

/******************************************************************************
 *
 *  Filename:      bt_vendor_nxp.h
 *
 *  Description:   NXP VHAL related declaration
 *
 ******************************************************************************/

#ifndef BT_VENDOR_NXP_H
#define BT_VENDOR_NXP_H

/*============================== Include Files ===============================*/

#include <fcntl.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <termios.h>
#include <time.h>
#include <unistd.h>

#include "bt_vendor_lib.h"
#include <stdbool.h>

/*================================== Macros ==================================*/

#define BT_HAL_VERSION "009.022"

#define TIMEOUT_SEC 6
#define RW_SUCCESSFUL (1)
#define RW_FAILURE (~RW_SUCCESSFUL)

#define TRUE 1
#define FALSE 0

#define DOWNLOAD_SUCCESS 0x0
#define OPEN_SERIAL_PORT_OR_FILE_ERROR 0x1
#define FEEK_SEEK_ERROR 0x2
#define FILESIZE_IS_ZERO 0x3
#define HEADER_SIGNATURE_TIMEOUT 0x4
#define READ_FILE_FAIL 0x5
#define CHANGE_BAUDRATE_FAIL 0x6
#define CHANGE_TIMEOUT_VALUE_FAIL 0x7
#define OPEN_FILE_FAIL 0x8
#define FILE_MODE_CANNOT_CHANGE 0X9
#define UNEXPECTED_BEHAVIOUR_IN_SETJMP 0xA
#define MALLOC_RETURNED_NULL 0xB
#define START_INDICATION_NOT_FOUND 0xC
#define INVALID_LEN_TO_SEND 0xD

#define BLE_SET_1M_POWER 0x01
#define BLE_SET_2M_POWER 0x02
#define BT_SET_SLEEP_MODE 0x02
#define BT_SET_FULL_POWER_MODE 0x03

#define HCI_CMD_INBAND_RESET 0xFCFC
#define HCI_CMD_NXP_RESET 0x0C03
#define HCI_CMD_NXP_CHANGE_BAUDRATE 0xFC09
#define HCI_CMD_NXP_BLE_WAKEUP 0xFD52
#define HCI_CMD_OTT_SUB_WAKEUP_EXIT_HEARTBEATS 0x08

#define WRITE_BD_ADDRESS_SIZE 8
#define MAX_PATH_LEN 512
#define MAX_DEVICE_LEN 32
#define MAX_FILE_LEN 128

#define HCI_EVENT_PAYLOAD_SIZE 255
#define MAX_CONF_PARA_LEN 36

/* BLUETOOTH CORE SPECIFICATION Version 5.4*/
/*(Volume 4, Part A, 2) | 1 byte HCI packet type*/
#define HCI_PACKET_TYPE_SIZE 1
/*(Volume 4,Part E, 5.4.4) 1 byte Event Code, 1 byte Parameter Total Length */
#define HCI_EVENT_HEADER_SIZE 2
/*(Volume 4, Part E, 5.4.1) | 2 bytes Opcode, 1 byte Parameter Total Length */
#define HCI_COMMAND_HEADER_SIZE 3

#define NXP_WAKEUP_ADV_PATTERN_LENGTH 16  // company id + vendor information
#define PROP_BLUETOOTH_INIT_ATTEMPTED "bluetooth.nxp.init_attempted"
#define PROP_VENDOR_TRIGGER_PDN "vendor.nxp.trigger_pdn"
#define PDN_RECOVERY_THRESHOLD (2)
#define PROP_BLUETOOTH_FW_DOWNLOADED "bluetooth.nxp.fw_downloaded"
#define PROP_BLUETOOTH_INBAND_CONFIGURED "bluetooth.nxp.inband_ir_configured"
#define PROP_BLUETOOTH_BOOT_SLEEP_TRIGGER \
  "bluetooth.nxp.sent_boot_sleep_triggered"
/* Run-time configuration file */
#ifndef VENDOR_LIB_CONF_FILE
#define VENDOR_LIB_CONF_FILE "/vendor/etc/bluetooth/bt_vendor.conf"
#endif
#define HCI_PACKET_COMMAND 0x01
#define HCI_PACKET_EVENT 0x04
#define HCI_EVENT_COMMAND_COMPLETE 0x0E
#define HCI_EVENT_HARDWARE_ERROR 0x10

#define HCI_EVT_PYLD_OPCODE_IDX 1
#define HCI_EVT_PYLD_STATUS_IDX 3
#define HCI_EVT_PYLD_SUBCODE_IDX 4

#define STREAM_TO_UINT16(u16, p)                                \
  do {                                                          \
    u16 = ((uint16_t)(*(p)) + (((uint16_t)(*((p) + 1))) << 8)); \
    (p) += 2;                                                   \
  } while (0)

#define UINT32_TO_STREAM(p, u32)     \
  do {                               \
    *(p)++ = (uint8_t)(u32);         \
    *(p)++ = (uint8_t)((u32) >> 8);  \
    *(p)++ = (uint8_t)((u32) >> 16); \
    *(p)++ = (uint8_t)((u32) >> 24); \
  } while (0)

#define UINT16_TO_STREAM(p, u16)    \
  do {                              \
    *(p)++ = (uint8_t)(u16);        \
    *(p)++ = (uint8_t)((u16) >> 8); \
  } while (0)

#define UINT8_TO_STREAM(p, u8) \
  { *(p)++ = (uint8_t)(u8); }
/*================================== Typedefs=================================*/

typedef unsigned long long uint64;
typedef unsigned int uint32;
typedef unsigned short uint16;
typedef unsigned char uint8;
typedef int int32;
typedef short int16;
typedef char int8;
typedef unsigned char BOOLEAN;

enum { IR_TRIGGER_NONE, IR_TRIGGER_RFKILL, IR_TRIGGER_GPIO };
enum { IR_MODE_NONE, IR_MODE_OOB_VSC, IR_MODE_INBAND_VSC };
enum { wakeup_power_key, wakeup_netflix_key, wakeup_key_num };

typedef struct {
  unsigned char gpio_pin;
  unsigned char high_duration;  // 100ms
  unsigned char low_duration;   // 100ms
} wakeup_gpio_config_t;

typedef struct {
  unsigned char length;
  unsigned char adv_pattern[NXP_WAKEUP_ADV_PATTERN_LENGTH];
} wakeup_adv_pattern_config_t;

typedef struct {
  unsigned char le_scan_type;
  unsigned short interval;
  unsigned short window;
  unsigned char own_addr_type;
  unsigned char scan_filter_policy;
} wakeup_scan_param_config_t;

typedef struct {
  unsigned char heartbeat_timer_value;  // 100ms
} wakeup_local_param_config_t;

typedef union {
  uint8_t raw_data[HCI_EVENT_PAYLOAD_SIZE + HCI_EVENT_HEADER_SIZE +
                   HCI_PACKET_TYPE_SIZE];
  struct info {
    uint8_t packet_type;
    uint8_t event_type;
    uint8_t para_len;
    uint8_t payload[];
  } info;
} hci_event;
/*================================ Global Vars================================*/

extern unsigned char* bdaddr;
extern int write_bdaddrss;
extern uint8_t write_bd_address[WRITE_BD_ADDRESS_SIZE];
extern const bt_vendor_callbacks_t* vnd_cb;
extern char pFilename_cal_data[MAX_PATH_LEN];
extern int8_t ble_1m_power;
extern int8_t ble_2m_power;
extern int8_t bt_max_power_sel;
extern uint8_t set_1m_2m_power;
extern uint8_t bt_set_max_power;
extern uint8_t independent_reset_mode;
extern uint8_t independent_reset_gpio_pin;
extern bool enable_sco_config;
extern bool use_controller_addr;
extern char pFilename_fw_init_config_bin[MAX_PATH_LEN];
extern bool enable_heartbeat_config;
extern bool enable_pdn_recovery;
extern wakeup_gpio_config_t wakeup_gpio_config[wakeup_key_num];
extern wakeup_adv_pattern_config_t wakeup_adv_config;
extern wakeup_scan_param_config_t wakeup_scan_param_config;
extern wakeup_local_param_config_t wakeup_local_param_config;
extern bool lpm_configured;
extern int mchar_fd;
extern bool wakeup_enable_uart_low_config;
int8 hw_bt_configure_lpm(uint8 sleep_mode);
void wakeup_kill_heartbeat_thread(void);
#ifdef UART_DOWNLOAD_FW
extern uint8_t enable_poke_controller;
#endif

/*============================ Function Prototypes ===========================*/

void hw_config_start(void);
int32 init_uart(int8* dev, int32 dwBaudRate, uint8 ucFlowCtrl);
int get_prop_int32(const char* name);
void set_prop_int32(const char* name, int value);
int8 hw_bt_send_wakeup_disable_raw(void);
int8 hw_bt_send_hci_cmd_raw(uint16_t opcode);
int8 hw_send_change_baudrate_raw(uint32_t baudrate);
const char* hw_bt_cmd_to_str(uint16_t cmd);
#endif  // BT_VENDOR_NXP_H
